devtools/server/actors/replay/connection.js

```diff
@@ -240,6 +281,322 @@ class Recording extends EventEmitter {
+  const key = getRecordingKey(browser);
+
+  if (isRecordingAllTabs()) {
+    return;
+  }
+
+  const tabbrowser = browser.getTabBrowser();
+  const tab = tabbrowser.getTabForBrowser(browser);
+  const contentPrincipal = browser.contentPrincipal;
+
+  const state = SessionStore.getTabState(tab);
+  tabbrowser.updateBrowserRemoteness(browser, {
+    recordExecution: undefined,
+    newFrameloader: true,
+    remoteType: E10SUtils.WEB_REMOTE_TYPE,
+  });
+
+  if (!url) {
+    const contentUrl = browser.currentURI.spec;
+
+    browser.loadURI(contentUrl, { triggeringPrincipal: contentPrincipal });
+  }
+
+  // Creating a new frameloader will destroy the tab's session history so we
+  // need to restore it.
+  SessionStore.setTabState(tab, state);
+
+  if (url) {
+    browser.loadURI(url, {
+      triggeringPrincipal: Services.scriptSecurityManager.getSystemPrincipal()
+    });
+  }
+
+  remapRecordingState(browser, key);
+}
+
+function setRecordingSaved(browser, recordingId) {
+  // suppress launching new tab for test recordings
+  if (Services.prefs.getBoolPref("devtools.recordreplay.submitTestRecordings")) {
+    return;
+  }
+
+  // Find the dispatcher to connect to.
+  const dispatchAddress = getDispatchServer();
+  const key = getRecordingKey(browser);
+
+  let extra = "";
+
+  // Specify the dispatch address if it is not the default.
+  if (dispatchAddress != "wss://dispatch.replay.io") {
+    extra += `&dispatch=${dispatchAddress}`;
+  }
+
+  // For testing, allow specifying a test script to load in the tab.
+  const localTest = getenv("RECORD_REPLAY_LOCAL_TEST");
+  if (localTest) {
+    extra += `&test=${localTest}`;
+  } else if (!isAuthenticationEnabled()) {
+    // Adding this urlparam disables checks in the devtools that the user has
+    // permission to view the recording.
+    extra += `&test=1`;
+  }
+
+  const tabbrowser = browser.getTabBrowser();
+  const currentTabIndex = tabbrowser.visibleTabs.indexOf(tabbrowser.selectedTab);
+  const triggeringPrincipal = Services.scriptSecurityManager.getSystemPrincipal();
+  const tab = tabbrowser.addTab(
+    `${getViewURL()}?id=${recordingId}${extra}`,
+    { triggeringPrincipal, index: currentTabIndex === -1 ? undefined : currentTabIndex + 1}
+  );
+  tabbrowser.selectedTab = tab;
+
+  // defer setting the state until the end so that the new tab has opened before the spinner disappears.
+  setRecordingState(key, RecordingState.READY);
+}
+
+function handleRecordingStarted(pmm) {
+  const recording = new Recording(pmm);
+
+  // There can occasionally be times when the browser isn't found when the
+  // recording begins, so we lazily look it up the first time it is needed.
+  let _browser = null;
+  function getBrowser() {
+    for (let frameLoader of recordings.keys()) {
+      if (frameLoader.remoteTab && frameLoader.remoteTab.osPid === pmm.osPid) {
+        _browser = frameLoader.ownerElement;
+        break;
+      }
+    }
+    return _browser;
+  }
```

Comment:
I'm surprised that this isn't failing inside `saved`. I'd have assumed that `remoteTab` would be gone by that point in time so it would fail to find the browser, since it will have been `remap`-ed at by then.

What I had in mind for this logic is more along the lines of this:
```
function handleRecordingStarted(pmm) {
  const recording = new Recording(pmm);

  let frameLoader = null;
  for (const window of Services.wm.getEnumerator("navigator:browser")) {
    for (const tab of window.gBrowser.tabs) {
      const { remoteTab } = tab.linkedBrowser.frameLoader || {};
      
      if (remoteTab && remoteTab.osPid === pmm.osPid) {
        frameLoader = tab.linkedBrowser.frameLoader;
		break;
      }
    }
  }
  if (!frameLoader) {
    throw new Error("Unable to find browser for recording");
  }

  const recording = new Recording(pmm, frameLoader);

  // ...
}
```
this mean that the `Recording` instance basically represents a fully in-progress recording and knows what frameLoader is the one associated with the recording, and instead of `state` being on a separate object, `state` could be part of the `Recording`.

That said, honestly I think this is getting more complicated than I want it to right now, so we should probably think about just leaving this as it was for now.

Reply:
> I'm surprised that this isn't failing inside saved. I'd have assumed that remoteTab would be gone by that point in time so it would fail to find the browser, since it will have been remap-ed at by then.

I suspect though I haven't tested that `saved` is using the cached browser from `finished`.

> this mean that the Recording instance basically represents a fully in-progress recording and knows what frameLoader is the one associated with the recording, and instead of state being on a separate object, state could be part of the Recording.

I considered consolidating state into Recording but we'd have to create it sooner and then populate it with `pmm` later because we need to start tracking the state as soon as the user clicks the button.

Reply:
> I suspect though I haven't tested that saved is using the cached browser from finished.

Ohh because we don't reset it to `null` before looping to search, _right_. Fair enough. 

> I considered consolidating state into Recording but we'd have to create it sooner and then populate it with pmm later because we need to start tracking the state as soon as the user clicks the button.

Ah right. That centralization seems worth it to me but not necessary as part of this patch. There's plenty more to be done here eventually anyway.